File: 160.intersection-of-two-linked-lists.js
/*
 * @lc app=leetcode id=160 lang=javascript
 *
 * [160] Intersection of Two Linked Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(nodeA, nodeB) {
  if (!nodeA || !nodeB) return null
  if (nodeA == nodeB) return nodeA
  let a = [], b = []

  do {
    if (nodeA) {
      a.push(nodeA)
      nodeA = nodeA.next
    }
    if (nodeB) {
      b.push(nodeB)
      nodeB = nodeB.next
    }
    if (a.includes(nodeB)) return nodeB
    if (b.includes(nodeA)) return nodeA
    if (nodeA === nodeB) return nodeA
  } while (nodeA || nodeB)
  return null
};
// @lc code=end

File: 160.intersection-of-two-linked-lists.py
#
# @lc app=leetcode id=160 lang=python3
#
# [160] Intersection of Two Linked Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        hash = {}
        while headA:
            if headA.val not in hash: hash[headA.val] = []
            hash[headA.val].append(headA)
            headA = headA.next
        while headB:
            if headB.val in hash and headB in hash[headB.val]:
                return headB
            headB = headB.next
        return None
# @lc code=end

